packages/plugin/src/graphql-ast.ts

```diff
@@ -128,3 +130,87 @@ export function collectReachableTypes(schema: GraphQLSchema): Set<string> {
     return false;
   }
 }
+
+export type FieldsCache = Record<string, Set<string>>;
+
+export function createUsedFieldsService(schema: GraphQLSchema, operations: SiblingOperations): () => FieldsCache | null {
+  if (!schema || !operations) {
+    return () => null;
+  }
+
+  let cache: FieldsCache = null;
+
+  return () => {
+    if (!cache) {
+      cache = collectUsedFields(schema, operations);
+    }
+
+    return cache;
+  };
+}
+
+export function collectUsedFields(schema: GraphQLSchema, operations: SiblingOperations): FieldsCache {
+  const typesByOperations = {
+    query: schema.getQueryType(),
+    mutation: schema.getMutationType(),
+    subscription: schema.getSubscriptionType(),
+  };
+
+  const cache: FieldsCache = {};
+
+  const addField = (typeName, fieldName) => {
+    const fieldType = cache[typeName] ?? (cache[typeName] = new Set());
+    fieldType.add(fieldName);
+  };
+
+  const visit = (selection, type?: GraphQLNamedType) => {
+    switch (selection.kind) {
+      case Kind.OPERATION_DEFINITION:
+        type = typesByOperations[selection.operation];
+        break;
+      case Kind.FRAGMENT_DEFINITION:
+      case Kind.INLINE_FRAGMENT:
+        type = schema.getType(selection.typeCondition.name.value);
```

Comment:
I think `typeCondition` could be `null`, we had an issue with that in codegen, worth adding the extra check.

Reply:
Same, after refactor I don't use this code anymore, but thanks 👍